### PR TITLE
feat(tcp): multithreaded server and client with promt

### PR DESCRIPTION
**Pull request by yuriykulikov**

### CODE REVIEW

File: tcp-server/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+use redisish::{Command, parse};
+
+trait Mailbox {
+    /// TODO Is it fine that mutability is hidden? How can we do it some other way?
+    fn append(&self, email: String);
```
Comment (spookyvision) [inline tcp-server/src/main.rs:17]: I'd make it explicit by declaring `append(&mut self`

File: tcp-server/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    /// TODO does it make sense to return an abstract iterator and how can this be achieved?
+    fn list_emails(&self) -> String;
```
Comment (spookyvision) [inline tcp-server/src/main.rs:19]: out of scope for this exercise; the [leveldb trail](https://exercises-2021.ferrous-systems.com/ffi-1-exercise.html) has a [section](https://exercises-2021.ferrous-systems.com/ffi-3-exercise.html) on iterators

File: tcp-server/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+struct VecDequeMailbox {
+    data: Mutex<VecDeque<String>>,
+}
+
+impl Mailbox for VecDequeMailbox {
+    /// TODO is there a way (and does it make sense) to accept &str?
+    fn append(&self, email: String) {
```
Comment (spookyvision) [inline tcp-server/src/main.rs:28]: it does make sense! I'd use `AsRef<str>` here and `a_str.as.ref()` / `an_owned_string.as_ref()` at the call sites

File: tcp-server/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    fn list_emails(&self) -> String {
+        self.data.lock()
+            .unwrap()
+            .iter()
+            .fold(String::new(), |acc, next| acc + next + ";")
+    }
+}
+
+impl VecDequeMailbox {
+    // TODO how to return Box<dyn Mailbox>?
```
Comment (spookyvision) [inline tcp-server/src/main.rs:43]: you have only one implementation, so I would not use traits + dyn here; generally speaking you should be able to just declare `Box<dyn Mailbox>` as return type. Ping me if you have issues with that!

File: tcp-server/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    fn new() -> Box<VecDequeMailbox> {
+        return Box::new(VecDequeMailbox {
+            data: Mutex::new(VecDeque::new()),
+        });
+    }
+}
+
+/// General questions:
+/// * how to work with `dyn Mailbox ` in multithreaded environment?
```
Comment (spookyvision) [inline tcp-server/src/main.rs:52]: you can further restrict your type by saying Box<dyn Mailbox + Send + Sync> for example. But I would not use traits like that in this exercise

File: tcp-server/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+/// Spawns TcpListener thread, connects to 127.0.0.1:8080 and spawns thread per connection
+fn spawn_tcp_listener() {
+    thread::spawn(|| {
+        let mailbox = Arc::new(VecDequeMailbox::new());
+        let listener = TcpListener::bind("127.0.0.1:8080").unwrap();
+        for stream in listener.incoming() {
+            // TODO is shadowing the variable name idiomatic?
+            let mailbox = mailbox.clone();
```
Comment (spookyvision) [inline tcp-server/src/main.rs:67]: yes!

File: tcp-server/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            // TODO: confirm that dropping the JoinHandle does not do anything
```
Comment (spookyvision) [inline tcp-server/src/main.rs:68]: thread handles should be joined for proper error handling but it's not a fatal error to just drop them. In this case you can't wait for them to finish anymore, so any chance of re-sync is lost. See [JoinHandle docs](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html)

File: tcp-server/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            thread::spawn(|| {
+                println!("Server has a new client {:?}", stream);
+                handle_client(stream.unwrap(), mailbox);
+            });
+        }
+    });
+}
+
+/// TODO how can I pass the mailbox, preferably just mailbox: Mailbox or at least mailbox: Arc<Box<dyn Mailbox>>?
```
Comment (spookyvision) [inline tcp-server/src/main.rs:77]: for accessing things from different threads you need synchronization primitives, which typically means a smart pointer to a mutex around your data: `Arc<Mutex<Mailbox>>`

File: tcp-server/src/main.rs (hunk 1 above)
Comment (yuriykulikov) [reply tcp-server/src/main.rs:17]: In order to achieve that I will need a lock around it, right? The struct is used from multiple threads/tasks/functions.

What would be preferable in this case, a lock around the struct or a lock inside the struct?

File: tcp-server/src/main.rs (hunk 3 above)
Comment (yuriykulikov) [reply tcp-server/src/main.rs:28]: Done

File: tcp-server/src/main.rs (hunk 4 above)
Comment (yuriykulikov) [reply tcp-server/src/main.rs:43]: Going with async functions means that Mailbox cannot be a trait for now. :(
My motivation was to make Mailbox an abstraction for multiple implementations,  for example, VecDequeMailbox, FileMailbox, FirebaseMailbox.

File: tcp-server/src/main.rs (hunk 6 above)
Comment (yuriykulikov) [reply tcp-server/src/main.rs:67]: Cool, thanks!

**Pull request merged**